In [8]:
import sys


In [9]:
import sys

print(sys.executable)
print(sys.version)

c:\Users\shafi\number-simplex-reproduction\.venv\Scripts\python.exe
3.14.2 (tags/v3.14.2:df79316, Dec  5 2025, 17:18:21) [MSC v.1944 64 bit (AMD64)]


In [10]:
from pathlib import Path

raw_dir = Path("../data/raw")

for subject_dir in sorted(raw_dir.iterdir()):
    if subject_dir.is_dir():
        print(f"\nSubject: {subject_dir.name}")

        for path in sorted(subject_dir.rglob("*")):
            if path.is_file():
                print("   ", path.relative_to(subject_dir))


Subject: YFF
    .DS_Store
    arithmetic\photoBehavEvents.csv
    arithmetic\photoBehavEvents.mat
    arithmetic\spikes.mat
    dot\photoBehavEvents.mat

Subject: YFI
    .DS_Store
    arithmetic\photoBehavEvents.mat
    arithmetic\spikes.mat
    dot\photoBehavEvents.mat
    dot\spikes.mat

Subject: YFJ
    .DS_Store
    arithmetic\photoBehavEvents.mat
    arithmetic\spikes.mat
    dot\photoBehavEvents.mat
    dot\spikes.mat

Subject: YFK
    .DS_Store
    arithmetic\photoBehavEvents.mat
    arithmetic\spikesArithmetic.mat
    dot\photoBehavEvents.mat
    dot\spikesDot.mat

Subject: YFL
    .DS_Store
    arithmetic\photoBehavEvents.mat
    arithmetic\spikesArithmetic.mat
    dot\photoBehavEvents.mat
    dot\spikesDot.mat

Subject: YFM
    .DS_Store
    arithmetic\photoBehavEvents.mat
    arithmetic\spikesArithmetic.mat
    dot\photoBehavEvents.mat
    dot\spikesDot.mat

Subject: YFP
    .DS_Store
    arithmetic\photoBehavEvents.mat
    arithmetic\spikes.mat

Subject: YFR
    .DS_Stor

In [25]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    print("Top-level keys:")
    for key in f.keys():
        print(key)
        

Top-level keys:
#refs#
regionsVect
spikes


In [26]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    print("Top-level keys:")
    for key in f.keys():
        print(key)

Top-level keys:
#refs#
regionsVect
spikes


In [27]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    for key in ["regionsVect", "spikes"]:
        obj = f[key]

        print(f"\n{key}")
        print("Type :", type(obj))
        print("Shape:", obj.shape)
        print("Dtype:", obj.dtype)


regionsVect
Type : <class 'h5py._hl.dataset.Dataset'>
Shape: (1, 54)
Dtype: object

spikes
Type : <class 'h5py._hl.group.Group'>


AttributeError: 'Group' object has no attribute 'shape'

In [ ]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    spikes_group = f["spikes"]

    print("Objects inside spikes:")
    
    for key in spikes_group.keys():
        obj = spikes_group[key]
        print(
            key,
            "| type:", type(obj),
            "| shape:", getattr(obj, "shape", None),
            "| dtype:", getattr(obj, "dtype", None)
        )
        

Objects inside spikes:
data | type: <class 'h5py._hl.dataset.Dataset'> | shape: (501537,) | dtype: uint8
ir | type: <class 'h5py._hl.dataset.Dataset'> | shape: (501537,) | dtype: uint64
jc | type: <class 'h5py._hl.dataset.Dataset'> | shape: (708277,) | dtype: uint64


In [ ]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:

    spikes = f["spikes"]

    print("SPIKES GROUP ATTRIBUTES")
    for key, value in spikes.attrs.items():
        print(key, ":", value)

    print("\nDATA ATTRIBUTES")
    for key, value in spikes["data"].attrs.items():
        print(key, ":", value)

    print("\nIR ATTRIBUTES")
    for key, value in spikes["ir"].attrs.items():
        print(key, ":", value)

    print("\nJC ATTRIBUTES")
    for key, value in spikes["jc"].attrs.items():
        print(key, ":", value)

SPIKES GROUP ATTRIBUTES
MATLAB_class : b'logical'
MATLAB_int_decode : 1
MATLAB_sparse : 54

DATA ATTRIBUTES

IR ATTRIBUTES

JC ATTRIBUTES


In [ ]:
import h5py
import numpy as np
from scipy.sparse import csc_matrix

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    spikes = f["spikes"]

    data = spikes["data"][:]
    ir = spikes["ir"][:]
    jc = spikes["jc"][:]

    n_rows = int(spikes.attrs["MATLAB_sparse"])
    n_cols = len(jc) - 1

spike_matrix = csc_matrix(
    (data, ir, jc),
    shape=(n_rows, n_cols)
)

print("Shape:", spike_matrix.shape)
print("Nonzero entries:", spike_matrix.nnz)
print("Data values:", np.unique(spike_matrix.data))

Shape: (54, 708276)
Nonzero entries: 501537
Data values: [1]


In [ ]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    regions = f["regionsVect"]

    print("Shape:", regions.shape)
    print("Dtype:", regions.dtype)

    refs = regions[0]

    for i, ref in enumerate(refs[:10]):
        obj = f[ref]
        print(i, obj[()])

Shape: (1, 54)
Dtype: object
0 [[97]
 [99]
 [99]]
1 [[97]
 [99]
 [99]]
2 [[97]
 [99]
 [99]]
3 [[97]
 [99]
 [99]]
4 [[97]
 [99]
 [99]]
5 [[97]
 [99]
 [99]]
6 [[97]
 [99]
 [99]]
7 [[97]
 [99]
 [99]]
8 [[101]
 [110]
 [116]]
9 [[104]
 [112]
 [ 99]]


In [ ]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    regions = f["regionsVect"]

    region_labels = []

    for ref in regions[0]:
        obj = f[ref]
        char_codes = obj[()].flatten()

        label = "".join(chr(int(c)) for c in char_codes)
        region_labels.append(label)

for i, label in enumerate(region_labels):
    print(f"Neuron {i:2d}: {label}")

Neuron  0: acc
Neuron  1: acc
Neuron  2: acc
Neuron  3: acc
Neuron  4: acc
Neuron  5: acc
Neuron  6: acc
Neuron  7: acc
Neuron  8: ent
Neuron  9: hpc
Neuron 10: hpc
Neuron 11: hpc
Neuron 12: hpc
Neuron 13: hpc
Neuron 14: hpc
Neuron 15: hpc
Neuron 16: hpc
Neuron 17: acc
Neuron 18: acc
Neuron 19: acc
Neuron 20: acc
Neuron 21: acc
Neuron 22: acc
Neuron 23: acc
Neuron 24: acc
Neuron 25: acc
Neuron 26: hpc
Neuron 27: hpc
Neuron 28: hpc
Neuron 29: hpc
Neuron 30: hpc
Neuron 31: hpc
Neuron 32: hpc
Neuron 33: hpc
Neuron 34: hpc
Neuron 35: hpc
Neuron 36: hpc
Neuron 37: hpc
Neuron 38: hpc
Neuron 39: hpc
Neuron 40: hpc
Neuron 41: hpc
Neuron 42: hpc
Neuron 43: hpc
Neuron 44: hpc
Neuron 45: hpc
Neuron 46: hpc
Neuron 47: hpc
Neuron 48: hpc
Neuron 49: hpc
Neuron 50: hpc
Neuron 51: hpc
Neuron 52: hpc
Neuron 53: hpc


In [ ]:
from collections import Counter

region_counts = Counter(region_labels)

print("Region counts:")
for region, count in sorted(region_counts.items()):
    print(f"{region}: {count}")

Region counts:
acc: 17
ent: 1
hpc: 36


In [ ]:
from collections import Counter

region_counts = Counter(region_labels)

print("Region counts:")
for region, count in sorted(region_counts.items()):
    print(f"{region}: {count}")

Region counts:
acc: 17
ent: 1
hpc: 36


In [ ]:
import h5py

file_path = "../data/raw/YFF/arithmetic/spikes.mat"

with h5py.File(file_path, "r") as f:
    print("FILE ATTRIBUTES:")
    for key, value in f.attrs.items():
        print(key, ":", value)

    print("\nSPIKES ATTRIBUTES:")
    for key, value in f["spikes"].attrs.items():
        print(key, ":", value)

    print("\nREGIONSVECT ATTRIBUTES:")
    for key, value in f["regionsVect"].attrs.items():
        print(key, ":", value)

FILE ATTRIBUTES:

SPIKES ATTRIBUTES:
MATLAB_class : b'logical'
MATLAB_int_decode : 1
MATLAB_sparse : 54

REGIONSVECT ATTRIBUTES:
MATLAB_class : b'cell'


In [16]:
import h5py

behav_path = "../data/raw/YFF/arithmetic/photoBehavEvents.mat"

with h5py.File(behav_path, "r") as f:
    print("Top-level keys:")

    for key in f.keys():
        obj = f[key]

        print(
            key,
            "| type:", type(obj),
            "| shape:", getattr(obj, "shape", None),
            "| dtype:", getattr(obj, "dtype", None)
        )

OSError: Unable to synchronously open file (file signature not found)

In [17]:
from scipy.io import loadmat

behav_path = "../data/raw/YFF/arithmetic/photoBehavEvents.mat"

behav = loadmat(behav_path)

print("Top-level keys:")
for key in behav.keys():
    print(key, type(behav[key]), getattr(behav[key], "shape", None))

Top-level keys:
__header__ <class 'bytes'> None
__version__ <class 'str'> None
__globals__ <class 'list'> None
photoBehavEvents <class 'scipy.io.matlab._mio5_params.MatlabOpaque'> (1,)
__function_workspace__ <class 'numpy.ndarray'> (1, 23024)


In [18]:
import pandas as pd

behav = pd.read_csv("../data/raw/YFF/arithmetic/photoBehavEvents.csv")

behav.head()

,trial,tCue1,tCue2,tCue3,presentationEnd,tPress1,tPress2,tPress3,keyPress1,keyPress2,...,timeFeedback,cue1,cue2,operation,operationFirst,correctAnsw,givenAnsw,correct,correctFb,toExclude
0,1,5338.733333,6089.466667,6840.200000,7607.600000,13830.233333,14030.500000,NaN,1,1.0,...,15398.433333,5,16,-,1,-11,11,0,1,0
1,2,15882.233333,16632.933333,17383.666667,18167.733333,20453.266667,NaN,NaN,3,NaN,...,21938.033333,2,1,+,0,3,3,1,1,0
2,3,22438.533333,23189.233333,23939.966667,24690.666667,26192.133333,NaN,NaN,7,NaN,...,27576.766667,6,1,+,1,7,7,1,1,0
3,4,28077.300000,28828.000000,29578.700000,30329.433333,34967.200000,35250.833333,NaN,2,1.0,...,36635.466667,0,16,+,0,16,21,0,1,0
4,5,37136.000000,37886.700000,38637.400000,39388.133333,40872.900000,41256.600000,NaN,1,6.0,...,42607.900000,10,6,+,0,16,16,1,1,0


In [19]:
behav.shape

(100, 22)

In [20]:
time_cols = [
    "tCue1",
    "tCue2",
    "tCue3",
    "presentationEnd",
    "tPress1",
    "tPress2",
    "tPress3",
    "choiceEnd",
    "timeFeedback",
]

behav[time_cols].max()

tCue1              701090.433333
tCue2              701824.433333
tCue3              702575.166667
presentationEnd    703325.866667
tPress1            705160.966667
tPress2            706328.766667
tPress3            544189.733333
choiceEnd          707129.533333
timeFeedback       707646.700000
dtype: float64

In [22]:
n_time_samples = spike_matrix.shape[1]

last_event_time = behav["timeFeedback"].max()

print("Spike recording length:", n_time_samples)
print("Last feedback time:", last_event_time)
print("Difference:", n_time_samples - last_event_time)

NameError: name 'spike_matrix' is not defined

In [23]:
trial1 = behav.iloc[0]

print("Cue1 → Cue2:", trial1["tCue2"] - trial1["tCue1"])
print("Cue2 → Cue3:", trial1["tCue3"] - trial1["tCue2"])

Cue1 → Cue2: 750.7333333333399
Cue2 → Cue3: 750.7333333333299


In [24]:
behav

,trial,tCue1,tCue2,tCue3,presentationEnd,tPress1,tPress2,tPress3,keyPress1,keyPress2,...,timeFeedback,cue1,cue2,operation,operationFirst,correctAnsw,givenAnsw,correct,correctFb,toExclude
0,1,5338.733333,6089.466667,6840.200000,7607.600000,13830.233333,14030.500000,NaN,1,1.0,...,15398.433333,5,16,-,1,-11,11,0,1,0
1,2,15882.233333,16632.933333,17383.666667,18167.733333,20453.266667,NaN,NaN,3,NaN,...,21938.033333,2,1,+,0,3,3,1,1,0
2,3,22438.533333,23189.233333,23939.966667,24690.666667,26192.133333,NaN,NaN,7,NaN,...,27576.766667,6,1,+,1,7,7,1,1,0
3,4,28077.300000,28828.000000,29578.700000,30329.433333,34967.200000,35250.833333,NaN,2,1.0,...,36635.466667,0,16,+,0,16,21,0,1,0
4,5,37136.000000,37886.700000,38637.400000,39388.133333,40872.900000,41256.600000,NaN,1,6.0,...,42607.900000,10,6,+,0,16,16,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,96,678001.566667,678752.266667,679503.000000,680237.033333,681554.966667,682038.766667,NaN,-,8.0,...,683406.733333,2,10,-,0,-8,-8,1,1,0
96,97,683907.233333,684657.933333,685408.666667,686159.400000,687794.300000,688127.966667,NaN,1,0.0,...,689379.133333,12,2,+,1,14,10,0,1,0
97,98,689879.633333,690613.666667,691381.066667,692115.100000,693082.700000,693399.700000,NaN,1,0.0,...,694784.333333,7,3,+,0,10,10,1,1,0
98,99,695268.166667,696018.866667,696786.266667,697537.000000,698771.500000,699205.266667,NaN,2,4.0,...,700573.233333,13,11,+,0,24,24,1,1,0


In [28]:
behav[:2]

,trial,tCue1,tCue2,tCue3,presentationEnd,tPress1,tPress2,tPress3,keyPress1,keyPress2,...,timeFeedback,cue1,cue2,operation,operationFirst,correctAnsw,givenAnsw,correct,correctFb,toExclude
0,1,5338.733333,6089.466667,6840.200000,7607.600000,13830.233333,14030.5,NaN,1,1.0,...,15398.433333,5,16,-,1,-11,11,0,1,0
1,2,15882.233333,16632.933333,17383.666667,18167.733333,20453.266667,NaN,NaN,3,NaN,...,21938.033333,2,1,+,0,3,3,1,1,0


In [30]:
6089.46-5338.73


750.7300000000005

In [35]:
behav[99:]

,trial,tCue1,tCue2,tCue3,presentationEnd,tPress1,tPress2,tPress3,keyPress1,keyPress2,...,timeFeedback,cue1,cue2,operation,operationFirst,correctAnsw,givenAnsw,correct,correctFb,toExclude
99,100,701090.433333,701824.433333,702575.166667,703325.866667,705160.966667,706328.766667,NaN,2,1.0,...,707646.7,5,16,+,0,21,21,1,1,0


In [37]:
701090 / 60000

11.684833333333334

Validation of each subject data

In [40]:
from pathlib import Path

raw_dir = Path("../data/raw")

subjects = sorted([p for p in raw_dir.iterdir() if p.is_dir()])

print("Number of subjects:", len(subjects))

for subject in subjects:
    print("\n", subject.name)

    for task in sorted([p for p in subject.iterdir() if p.is_dir()]):
        files = sorted([f.name for f in task.iterdir() if f.is_file()])
        print(f"  {task.name}: {files}")

Number of subjects: 11

 YFF
  arithmetic: ['photoBehavEvents.csv', 'photoBehavEvents.mat', 'spikes.mat']
  dot: ['photoBehavEvents.mat']

 YFI
  arithmetic: ['photoBehavEvents.mat', 'spikes.mat']
  dot: ['photoBehavEvents.mat', 'spikes.mat']

 YFJ
  arithmetic: ['photoBehavEvents.mat', 'spikes.mat']
  dot: ['photoBehavEvents.mat', 'spikes.mat']

 YFK
  arithmetic: ['photoBehavEvents.mat', 'spikesArithmetic.mat']
  dot: ['photoBehavEvents.mat', 'spikesDot.mat']

 YFL
  arithmetic: ['photoBehavEvents.mat', 'spikesArithmetic.mat']
  dot: ['photoBehavEvents.mat', 'spikesDot.mat']

 YFM
  arithmetic: ['photoBehavEvents.mat', 'spikesArithmetic.mat']
  dot: ['photoBehavEvents.mat', 'spikesDot.mat']

 YFP
  arithmetic: ['photoBehavEvents.mat', 'spikes.mat']
  dot: []

 YFR
  arithmetic: ['photoBehavEvents.mat', 'spikes.mat']
  dot: ['photoBehavEvents.mat', 'spikes.mat']

 YFS
  arithmetic: ['photoBehavEvents.mat', 'spikes.mat']
  dot: ['photoBehavEvents.mat', 'spikes.mat']

 YFT
  arithmetic:

In [42]:
import h5py
import pandas as pd

records = []

for subject in subjects:

    arithmetic_dir = subject / "arithmetic"

    # Handle the two different spike-file naming conventions
    possible_files = [
        arithmetic_dir / "spikes.mat",
        arithmetic_dir / "spikesArithmetic.mat"
    ]

    spike_file = next(
        (f for f in possible_files if f.exists()),
        None
    )

    if spike_file is None:
        records.append({
            "subject": subject.name,
            "neurons": None,
            "recording_ms": None,
            "recording_min": None
        })
        continue

    with h5py.File(spike_file, "r") as f:

        spikes = f["spikes"]

        n_neurons = int(spikes.attrs["MATLAB_sparse"])
        n_timepoints = len(spikes["jc"]) - 1

    records.append({
        "subject": subject.name,
        "neurons": n_neurons,
        "recording_ms": n_timepoints,
        "recording_min": n_timepoints / 1000 / 60
    })


arithmetic_summary = pd.DataFrame(records)

arithmetic_summary

KeyError: "Unable to synchronously open object (object 'spikes' doesn't exist)"

In [43]:
import h5py

for subject in subjects:
    arithmetic_dir = subject / "arithmetic"

    possible_files = [
        arithmetic_dir / "spikes.mat",
        arithmetic_dir / "spikesArithmetic.mat"
    ]

    spike_file = next((f for f in possible_files if f.exists()), None)

    if spike_file is None:
        print(subject.name, "-> no arithmetic spike file")
        continue

    with h5py.File(spike_file, "r") as f:
        print(
            subject.name,
            "->",
            spike_file.name,
            "-> keys:",
            list(f.keys())
        )

YFF -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFI -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFJ -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFK -> spikesArithmetic.mat -> keys: ['#refs#', 'regionsVect', 'spikesArithmetic']
YFL -> spikesArithmetic.mat -> keys: ['#refs#', 'regionsVect', 'spikesArithmetic']
YFM -> spikesArithmetic.mat -> keys: ['#refs#', 'regionsVect', 'spikesArithmetic']
YFP -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFR -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFS -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFT -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']
YFU -> spikes.mat -> keys: ['#refs#', 'regionsVect', 'spikes']


In [45]:

import h5py
import pandas as pd

records = []

for subject in subjects:

    arithmetic_dir = subject / "arithmetic"

    # Find arithmetic spike file
    if (arithmetic_dir / "spikes.mat").exists():
        spike_file = arithmetic_dir / "spikes.mat"
        spike_key = "spikes"

    elif (arithmetic_dir / "spikesArithmetic.mat").exists():
        spike_file = arithmetic_dir / "spikesArithmetic.mat"
        spike_key = "spikesArithmetic"

    else:
        records.append({
            "subject": subject.name,
            "neurons": None,
            "recording_ms": None,
            "recording_min": None
        })
        continue

    with h5py.File(spike_file, "r") as f:

        spikes = f[spike_key]

        n_neurons = int(spikes.attrs["MATLAB_sparse"])
        n_timepoints = len(spikes["jc"]) - 1

    records.append({
        "subject": subject.name,
        "neurons": n_neurons,
        "recording_ms": n_timepoints,
        "recording_min": n_timepoints / 1000 / 60
    })


arithmetic_summary = pd.DataFrame(records)

arithmetic_summary

,subject,neurons,recording_ms,recording_min
0,YFF,54,708276,11.804600
1,YFI,38,944701,15.745017
2,YFJ,62,4016441,66.940683
3,YFK,68,626723,10.445383
4,YFL,94,937166,15.619433
5,YFM,94,808864,13.481067
6,YFP,98,1798543,29.975717
7,YFR,78,3075950,51.265833
8,YFS,96,2215511,36.925183
9,YFT,89,3167222,52.787033


In [46]:
a =arithmetic_summary

In [54]:
(a.iloc[:, 1].sum())


np.int64(849)

In [55]:
print("Total neurons:", arithmetic_summary["neurons"].sum())

print(
    "Min neurons:",
    arithmetic_summary["neurons"].min()
)

print(
    "Max neurons:",
    arithmetic_summary["neurons"].max()
)


Total neurons: 849
Min neurons: 38
Max neurons: 98


In [56]:
import pandas as pd

validation_rows = []

for _, row in arithmetic_summary.iterrows():
    subject = row["subject"]

    csv_path = raw_dir / subject / "arithmetic" / "photoBehavEvents.csv"

    behav_subj = pd.read_csv(csv_path)

    last_feedback = behav_subj["timeFeedback"].max()
    recording_ms = row["recording_ms"]

    validation_rows.append({
        "subject": subject,
        "recording_ms": recording_ms,
        "last_feedback_ms": last_feedback,
        "difference_ms": recording_ms - last_feedback,
        "n_trials": len(behav_subj)
    })

time_validation = pd.DataFrame(validation_rows)

time_validation


,subject,recording_ms,last_feedback_ms,difference_ms,n_trials
0,YFF,708276,7.076467e+05,6.293000e+02,100
1,YFI,944701,9.440346e+05,6.664333e+02,100
2,YFJ,4016441,4.015801e+06,6.399000e+02,100
3,YFK,626723,6.258546e+05,8.684000e+02,100
4,YFL,937166,9.363126e+05,8.533667e+02,100
5,YFM,808864,8.080116e+05,8.523667e+02,100
6,YFP,1798543,8.484654e+05,9.500776e+05,141
7,YFR,3075950,1.528740e+06,1.547210e+06,160
8,YFS,2215511,7.941228e+05,1.421388e+06,160
9,YFT,3167222,1.685489e+06,1.481733e+06,300


In [57]:
from collections import Counter

def decode_matlab_string(f, ref):
    """Decode one MATLAB cell-string referenced through HDF5."""
    obj = f[ref]
    values = obj[()].flatten()
    return "".join(chr(int(x)) for x in values)


region_rows = []

for subject in subjects:

    arithmetic_dir = subject / "arithmetic"

    if (arithmetic_dir / "spikes.mat").exists():
        spike_file = arithmetic_dir / "spikes.mat"

    elif (arithmetic_dir / "spikesArithmetic.mat").exists():
        spike_file = arithmetic_dir / "spikesArithmetic.mat"

    else:
        continue

    with h5py.File(spike_file, "r") as f:

        regions = f["regionsVect"]
        refs = regions[0]

        region_labels = [
            decode_matlab_string(f, ref)
            for ref in refs
        ]

        counts = Counter(region_labels)

        region_rows.append({
            "subject": subject.name,
            "total": len(region_labels),
            **counts
        })


region_summary = pd.DataFrame(region_rows).fillna(0)

region_summary

,subject,total,acc,ent,hpc,OFC/vmPFC,ofc,thalamus,para-hpc,amy,thal
0,YFF,54,17.0,1.0,36,0.0,0.0,0.0,0.0,0.0,0.0
1,YFI,38,9.0,8.0,21,0.0,0.0,0.0,0.0,0.0,0.0
2,YFJ,62,0.0,0.0,45,17.0,0.0,0.0,0.0,0.0,0.0
3,YFK,68,5.0,0.0,26,0.0,9.0,10.0,18.0,0.0,0.0
4,YFL,94,28.0,15.0,42,0.0,0.0,9.0,0.0,0.0,0.0
5,YFM,94,0.0,29.0,32,0.0,26.0,7.0,0.0,0.0,0.0
6,YFP,98,28.0,0.0,43,0.0,27.0,0.0,0.0,0.0,0.0
7,YFR,78,13.0,0.0,41,0.0,0.0,0.0,0.0,24.0,0.0
8,YFS,96,17.0,0.0,37,0.0,9.0,0.0,0.0,22.0,11.0
9,YFT,89,9.0,17.0,27,0.0,11.0,0.0,0.0,8.0,17.0


In [58]:
region_summary["hpc"].sum()

np.int64(389)

In [61]:
region_summary.sum(numeric_only=True)

total        849.0
acc          142.0
ent           70.0
hpc          389.0
OFC/vmPFC     17.0
ofc           82.0
thalamus      26.0
para-hpc      18.0
amy           77.0
thal          28.0
dtype: float64

In [62]:
mtl_regions = ["hpc", "ent", "amy", "para-hpc"]

mtl_totals = region_summary[mtl_regions].sum()

print("MTL region totals:")
print(mtl_totals)

print("\nTotal MTL neurons:", int(mtl_totals.sum()))

print(
    "Non-MTL neurons:",
    int(region_summary["total"].sum() - mtl_totals.sum())
)

MTL region totals:
hpc         389.0
ent          70.0
amy          77.0
para-hpc     18.0
dtype: float64

Total MTL neurons: 554
Non-MTL neurons: 295


### Validation checkpoint: MTL population

Across the 11 arithmetic datasets:

- Total units in supplied spike files: 849
- Hippocampus (HPC): 389
- Entorhinal cortex (ENT): 70
- Amygdala (AMY): 77
- Parahippocampal cortex (para-HPC): 18
- Total MTL units: 554
- Non-MTL units: 295

The 554 MTL units exactly reproduce the population reported in the paper.

In [63]:
behavior_rows = []

for subject in subjects:
    csv_path = raw_dir / subject.name / "arithmetic" / "photoBehavEvents.csv"

    behav_subj = pd.read_csv(csv_path)

    behavior_rows.append({
        "subject": subject.name,
        "n_trials": len(behav_subj),
        "n_excluded": int(behav_subj["toExclude"].sum()),
        "n_correct": int(behav_subj["correct"].sum()),
        "accuracy": behav_subj["correct"].mean(),
        "n_addition": int((behav_subj["operation"] == "+").sum()),
        "n_subtraction": int((behav_subj["operation"] == "-").sum()),
        "operationFirst_1": int((behav_subj["operationFirst"] == 1).sum()),
        "operationFirst_0": int((behav_subj["operationFirst"] == 0).sum()),
        "cue1_min": behav_subj["cue1"].min(),
        "cue1_max": behav_subj["cue1"].max(),
        "cue2_min": behav_subj["cue2"].min(),
        "cue2_max": behav_subj["cue2"].max(),
    })

behavior_summary = pd.DataFrame(behavior_rows)

behavior_summary

,subject,n_trials,n_excluded,n_correct,accuracy,n_addition,n_subtraction,operationFirst_1,operationFirst_0,cue1_min,cue1_max,cue2_min,cue2_max
0,YFF,100,0,84,0.840000,51,49,43,57,0,16,0,16
1,YFI,100,1,87,0.870000,56,44,47,53,0,16,0,16
2,YFJ,100,1,65,0.650000,56,44,45,55,0,16,0,16
3,YFK,100,0,88,0.880000,56,44,45,55,0,16,0,16
4,YFL,100,0,76,0.760000,56,44,45,55,0,16,0,16
5,YFM,100,0,90,0.900000,56,44,45,55,0,16,0,16
6,YFP,141,0,105,0.744681,76,65,63,78,0,16,0,16
7,YFR,160,0,156,0.975000,160,0,0,160,0,10,0,10
8,YFS,160,0,152,0.950000,160,0,0,160,0,10,0,10
9,YFT,300,0,280,0.933333,154,146,145,155,0,10,0,10


In [64]:
# Load YFF arithmetic behavior
yff = pd.read_csv(
    "../data/raw/YFF/arithmetic/photoBehavEvents.csv"
)

cols = [
    "trial",
    "cue1",
    "cue2",
    "operation",
    "operationFirst",
    "tCue1",
    "tCue2",
    "tCue3",
    "correctAnsw",
    "givenAnsw",
    "correct"
]

yff[cols].head(10)

,trial,cue1,cue2,operation,operationFirst,tCue1,tCue2,tCue3,correctAnsw,givenAnsw,correct
0,1,5,16,-,1,5338.733333,6089.466667,6840.200000,-11,11,0
1,2,2,1,+,0,15882.233333,16632.933333,17383.666667,3,3,1
2,3,6,1,+,1,22438.533333,23189.233333,23939.966667,7,7,1
3,4,0,16,+,0,28077.300000,28828.000000,29578.700000,16,21,0
4,5,10,6,+,0,37136.000000,37886.700000,38637.400000,16,16,1
5,6,9,8,-,0,43108.400000,43859.100000,44609.800000,1,1,1
6,7,5,15,-,1,49247.600000,49998.333333,50749.033333,-10,10,0
7,8,8,1,+,0,58206.233333,58956.933333,59707.633333,9,9,1
8,9,9,1,+,0,64245.366667,64996.066667,65746.766667,10,10,1
9,10,0,16,-,0,70885.066667,71635.766667,72386.500000,-16,16,0


In [65]:
arithmetic_checks = []

for subject in subjects:
    csv_path = raw_dir / subject.name / "arithmetic" / "photoBehavEvents.csv"
    df = pd.read_csv(csv_path)

    # Calculate what the answer SHOULD be from cue1, operation, cue2
    calculated = df["cue1"].copy()

    plus_mask = df["operation"] == "+"
    minus_mask = df["operation"] == "-"

    calculated.loc[plus_mask] = (
        df.loc[plus_mask, "cue1"] +
        df.loc[plus_mask, "cue2"]
    )

    calculated.loc[minus_mask] = (
        df.loc[minus_mask, "cue1"] -
        df.loc[minus_mask, "cue2"]
    )

    matches = calculated == df["correctAnsw"]

    arithmetic_checks.append({
        "subject": subject.name,
        "trials": len(df),
        "matching": int(matches.sum()),
        "not_matching": int((~matches).sum())
    })

arithmetic_validation = pd.DataFrame(arithmetic_checks)

arithmetic_validation

,subject,trials,matching,not_matching
0,YFF,100,100,0
1,YFI,100,100,0
2,YFJ,100,100,0
3,YFK,100,100,0
4,YFL,100,100,0
5,YFM,100,100,0
6,YFP,141,141,0
7,YFR,160,160,0
8,YFS,160,160,0
9,YFT,300,300,0


In [66]:
yff_aligned = yff.copy()

# Create empty columns
yff_aligned["operand1_time"] = float("nan")
yff_aligned["operation_time"] = float("nan")
yff_aligned["operand2_time"] = float("nan")

# operationFirst = 0
mask0 = yff_aligned["operationFirst"] == 0

yff_aligned.loc[mask0, "operand1_time"] = yff_aligned.loc[mask0, "tCue1"]
yff_aligned.loc[mask0, "operand2_time"] = yff_aligned.loc[mask0, "tCue2"]
yff_aligned.loc[mask0, "operation_time"] = yff_aligned.loc[mask0, "tCue3"]

# operationFirst = 1
mask1 = yff_aligned["operationFirst"] == 1

yff_aligned.loc[mask1, "operation_time"] = yff_aligned.loc[mask1, "tCue1"]
yff_aligned.loc[mask1, "operand1_time"] = yff_aligned.loc[mask1, "tCue2"]
yff_aligned.loc[mask1, "operand2_time"] = yff_aligned.loc[mask1, "tCue3"]


In [67]:
yff_aligned[
    [
        "trial",
        "cue1",
        "operation",
        "cue2",
        "operationFirst",
        "operand1_time",
        "operation_time",
        "operand2_time"
    ]
].head(10)


,trial,cue1,operation,cue2,operationFirst,operand1_time,operation_time,operand2_time
0,1,5,-,16,1,6089.466667,5338.733333,6840.200000
1,2,2,+,1,0,15882.233333,17383.666667,16632.933333
2,3,6,+,1,1,23189.233333,22438.533333,23939.966667
3,4,0,+,16,0,28077.300000,29578.700000,28828.000000
4,5,10,+,6,0,37136.000000,38637.400000,37886.700000
5,6,9,-,8,0,43108.400000,44609.800000,43859.100000
6,7,5,-,15,1,49998.333333,49247.600000,50749.033333
7,8,8,+,1,0,58206.233333,59707.633333,58956.933333
8,9,9,+,1,0,64245.366667,65746.766667,64996.066667
9,10,0,-,16,0,70885.066667,72386.500000,71635.766667


In [68]:
for subj in ["YFR", "YFS"]:

    path = raw_dir / subj / "arithmetic" / "photoBehavEvents.csv"
    df = pd.read_csv(path)

    print(f"\n{subj}")

    display(
        df[
            [
                "trial",
                "cue1",
                "operation",
                "cue2",
                "operationFirst",
                "tCue1",
                "tCue2",
                "tCue3",
                "correctAnsw"
            ]
        ].head(5)
    )


YFR


,trial,cue1,operation,cue2,operationFirst,tCue1,tCue2,tCue3,correctAnsw
0,1,9,+,1,0,21758.966667,22492.266667,23225.600000,10
1,2,10,+,5,0,85674.900000,86424.866667,87174.866667,15
2,3,10,+,7,0,103741.333333,104491.333333,105241.333333,17
3,4,1,+,9,0,114491.200000,115241.200000,115991.200000,10
4,5,5,+,4,0,122591.100000,123341.100000,124091.100000,9



YFS


,trial,cue1,operation,cue2,operationFirst,tCue1,tCue2,tCue3,correctAnsw
0,1,6,+,1,0,16352.200000,17102.166667,17852.166667,7
1,2,3,+,3,0,29485.333333,30235.300000,30985.300000,6
2,3,0,+,0,0,33935.233333,34685.233333,35435.233333,0
3,4,7,+,9,0,39018.500000,39768.466667,40518.466667,16
4,5,8,+,10,0,46751.700000,47501.700000,48251.666667,18


In [69]:
def align_arithmetic_events(df, subject):
    df = df.copy()

    df["operand1_time"] = float("nan")
    df["operation_time"] = float("nan")
    df["operand2_time"] = float("nan")

    # Special protocol: YFR and YFS
    if subject in ["YFR", "YFS"]:

        df["operand1_time"] = df["tCue1"]
        df["operation_time"] = df["tCue2"]
        df["operand2_time"] = df["tCue3"]

    # Standard protocol
    else:

        # Operand1 -> Operand2 -> Operation
        mask0 = df["operationFirst"] == 0

        df.loc[mask0, "operand1_time"] = df.loc[mask0, "tCue1"]
        df.loc[mask0, "operand2_time"] = df.loc[mask0, "tCue2"]
        df.loc[mask0, "operation_time"] = df.loc[mask0, "tCue3"]

        # Operation -> Operand1 -> Operand2
        mask1 = df["operationFirst"] == 1

        df.loc[mask1, "operation_time"] = df.loc[mask1, "tCue1"]
        df.loc[mask1, "operand1_time"] = df.loc[mask1, "tCue2"]
        df.loc[mask1, "operand2_time"] = df.loc[mask1, "tCue3"]

    return df



In [71]:
yfr = pd.read_csv(
    "../data/raw/YFR/arithmetic/photoBehavEvents.csv"
)

yfr_aligned = align_arithmetic_events(yfr, "YFR")

yfr_aligned[
    [
        "trial",
        "cue1",
        "operation",
        "cue2",
        "operand1_time",
        "operation_time",
        "operand2_time"
    ]
].head()

,trial,cue1,operation,cue2,operand1_time,operation_time,operand2_time
0,1,9,+,1,21758.966667,22492.266667,23225.600000
1,2,10,+,5,85674.900000,86424.866667,87174.866667
2,3,10,+,7,103741.333333,104491.333333,105241.333333
3,4,1,+,9,114491.200000,115241.200000,115991.200000
4,5,5,+,4,122591.100000,123341.100000,124091.100000


In [73]:
print(spike_matrix.shape)

NameError: name 'spike_matrix' is not defined

In [74]:
trial = yff_aligned.iloc[0]

operand1_time = trial["operand1_time"]

start_ms = int(round(operand1_time + 50))
end_ms   = int(round(operand1_time + 950))

neuron_id = 0

spikes_window = spike_matrix[
    neuron_id,
    start_ms:end_ms
]

print("Trial:", trial["trial"])
print("Operand 1:", trial["cue1"])
print("Operand 1 onset:", operand1_time, "ms")

print("\nAnalysis window:")
print("Start:", start_ms, "ms")
print("End:", end_ms, "ms")
print("Length:", end_ms - start_ms, "ms")

print("\nNeuron:", neuron_id)
print("Number of spikes:", spikes_window.sum())

NameError: name 'spike_matrix' is not defined